File: sdk/runanywhere-android/core/src/main/kotlin/com/runanywhere/sdk/components/vad/VADServiceProvider.kt
package com.runanywhere.sdk.components.vad

import com.runanywhere.sdk.components.base.ModuleRegistry
import com.runanywhere.sdk.components.base.VADServiceProvider
import com.runanywhere.sdk.foundation.SDKLogger

/**
 * WebRTC VAD Service Provider
 */
class WebRTCVADServiceProvider : VADServiceProvider {
    private val logger = SDKLogger("WebRTCVADServiceProvider")

    override val name: String = "WebRTCVAD"

    override suspend fun createVADService(configuration: VADConfiguration): VADService {
        logger.info("Creating WebRTC VAD Service")
        val service = WebRTCVADService()
        service.initialize(configuration)
        return service
    }

    override fun canHandle(modelId: String?): Boolean {
        // WebRTC VAD doesn't use models, so it can handle any request
        return true
    }

    companion object {
        /**
         * Register this provider with the module registry
         */
        fun register() {
            ModuleRegistry.registerVADProvider(WebRTCVADServiceProvider())
        }
    }
}

File: sdk/runanywhere-swift/Sources/RunAnywhere/Capabilities/VAD/Protocol/VADServiceProvider.swift
//
//  VADServiceProvider.swift
//  RunAnywhere SDK
//
//  Protocol for VAD service providers (plugins)
//

import Foundation

/// Protocol for VAD service providers (plugins)
/// Allows registration of different VAD implementations
public protocol VADServiceProvider {

    /// Provider priority (higher = preferred)
    var priority: Int { get }

    /// Provider name for logging
    var name: String { get }

    /// Check if this provider can handle the given configuration
    /// - Parameter configuration: The VAD configuration
    /// - Returns: Whether this provider supports the configuration
    func canHandle(configuration: VADConfiguration) -> Bool

    /// Create a VAD service instance
    /// - Parameter configuration: The configuration for the service
    /// - Returns: A configured VAD service
    func createVADService(configuration: VADConfiguration) async throws -> VADService
}

// MARK: - Default Implementation

extension VADServiceProvider {
    public var priority: Int { 0 }

    public func canHandle(configuration: VADConfiguration) -> Bool {
        // Default: can handle any configuration
        return true
    }
}
